src/registry.rs

```diff
@@ -83,47 +124,47 @@ impl Registry {
         self.r.write().unwrap().unregister(c)
     }
 
-    pub fn write_pb<T: Write>(&self, _: &mut T) -> Result<()> {
-        Ok(())
+    fn gather(&self) -> Vec<proto::MetricFamily> {
+        self.r.read().unwrap().gather()
     }
 
-    pub fn write_text<T: Write>(&self, _: &mut T) -> Result<()> {
-        Ok(())
-    }
-}
+    pub fn scrap(&self, writer: &mut Write, encoder: &Encoder) -> Result<usize> {
```

Comment:
I think we should keep writer and encoder separate. 

Wrapping a writer, we may have problems:
 - How does it embed in libraries that pass a writer for every request? (Creating encoders every time.)
 - How to define the `Encode` trait and implement it?
     
   ```rust
trait EncoderA {
    fn writer(&mut self) -> &mut Write;
    fn encode(&mut self, Vec<Metric>) -> Result<usize>; 
}

trait EncoderB: Write {
    fn encode(&mut self, Vec<Metric>) -> Result<usize>;
}
```
   - EncoderA: Writting a custom encoder is hard, because the typical writer is `&mut Write`, so custom encoder must deal with lifetime.
   - EncoderB: Writting a custom encoder is impracticable when we have writer `&mut Write`.